core/src/main/java/org/apache/iceberg/io/datafile/WriteBuilder.java

```diff
@@ -0,0 +1,340 @@
+    return this;
+  }
+
+  /**
+   * Adds the new properties to the writer configuration.
+   *
+   * <p>Write configuration affects writer behavior. To add file metadata properties, use {@link
+   * #meta(String, String)} or {@link #meta(Map)}.
+   *
+   * @param properties a map of writer config properties
+   * @return this for method chaining
+   */
+  public WriteBuilder<A, E> set(Map<String, String> properties) {
+    properties.forEach(appenderBuilder::set);
+    return this;
+  }
+
+  /**
+   * Set a file metadata property.
+   *
+   * <p>Metadata properties are written into file metadata. To alter a writer configuration
+   * property, use {@link #set(String, String)} or {@link #set(Map)}.
+   *
+   * @param property a file metadata property name
+   * @param value config value
+   * @return this for method chaining
+   */
+  public WriteBuilder<A, E> meta(String property, String value) {
+    appenderBuilder.meta(property, value);
+    return this;
+  }
+
+  /**
+   * Add the new properties to file metadata.
+   *
+   * <p>Metadata properties are written into file metadata. To alter a writer configuration
+   * property, use {@link #set(String, String)}.
+   *
+   * @param properties a map of file metadata properties
+   * @return this for method chaining
+   */
+  public WriteBuilder<A, E> meta(Map<String, String> properties) {
+    properties.forEach(appenderBuilder::meta);
+    return this;
+  }
+
+  /** Sets the metrics configuration used for collecting column metrics for the created file. */
+  public WriteBuilder<A, E> metricsConfig(MetricsConfig newMetricsConfig) {
+    appenderBuilder.metricsConfig(newMetricsConfig);
+    return this;
+  }
+
+  /** Overwrite the file if it already exists. */
+  public WriteBuilder<A, E> overwrite() {
+    return overwrite(true);
+  }
+
+  /** Sets the overwrite flag. The default value is <code>false</code>. */
+  public WriteBuilder<A, E> overwrite(boolean enabled) {
+    appenderBuilder.overwrite(enabled);
+    return this;
+  }
+
+  /**
+   * Sets the encryption key used for writing the file. If encryption is not supported by the writer
+   * then an exception should be thrown.
+   */
+  public WriteBuilder<A, E> fileEncryptionKey(ByteBuffer encryptionKey) {
+    appenderBuilder.fileEncryptionKey(encryptionKey);
+    return this;
+  }
+
+  /**
+   * Sets the additional authentication data prefix used for writing the file. If encryption is not
+   * supported by the writer then an exception should be thrown.
+   */
+  public WriteBuilder<A, E> aADPrefix(ByteBuffer aadPrefix) {
+    appenderBuilder.aADPrefix(aadPrefix);
+    return this;
+  }
+
+  /** Sets the row schema for the delete writers. */
+  public WriteBuilder<A, E> withRowSchema(Schema newSchema) {
+    this.rowSchema = newSchema;
+    return this;
+  }
+
+  /** Sets the equality field ids for the equality delete writer. */
+  public WriteBuilder<A, E> withEqualityFieldIds(List<Integer> fieldIds) {
```

Comment:
What is the relationship between this builder and the builder for data or delete files? Do we need to pass this here?

I think the latest changes may have gone too far in combining interfaces. I think the object models should handle appender builders and those builders should be used by a data file builders and delete file builders that are shared. Combining the delete and data file builders creates confusing options, like this for a data file builder.